sklearn/cross_validation.py

```diff
@@ -332,6 +332,119 @@ def __len__(self):
         return self.n_folds
 
 
+def disjoint_label_folds(y, n_folds=3):
+    """Creates folds where a same label is not in two different folds.
+    
+    Parameters
+    ----------
+    y : iterable of shape (n_samples,)
+        Contains an id for each sample.
+        The folds are built so that the same id doesn't appear in two different folds.
+    
+    n_folds: int, default is 3
+        Number of folds to split the data into.
+        
+    Returns
+    -------
+    folds: numpy array of shape (n_samples, )
+        Array of integers between 0 and (n_folds - 1).
+        Folds[i] contains the folds to which sample i is assigned.
+        
+    Notes
+    -----
+    The folds are built by distributing the labels by frequency of appearance.
+    """
+    labels = np.array(y)
+    unique_labels, y = np.unique(labels, return_inverse=True)
+    
+    # number of occurrence of each label (its "weight")
+    weight_per_label = np.bincount(y)
+    # We want to distribute the heaviest weights first
+    ind = np.argsort(weight_per_label)[::-1]
+    weight_per_label = weight_per_label[ind]
```

Comment:
btw, why don't you call it sample_per_label instead of weight_per_label? I'm not sure what taking about weights adds.